docs/source/python/dataset.rst

```diff
@@ -340,6 +340,30 @@ when constructing a directory partitioning:
 Directory partitioning also supports providing a full schema rather than inferring
 types from file paths.
 
+Automatic partitioning detection
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If the directory is partitioned using the hive partitioning scheme (see above)
+then pyarrow will be able to automatically recognize the partitioning and include
+the partitioning information as a column in the returned table.  There is no
+need to specify the partitioning unless you need to override the inferred data
+types of the partitioning columns:
+
+.. code-block:: python
+
+    dataset = ds.dataset("hive_partitioned", format="parquet")
```

Comment:
To add to the confusion, on the issue you mentioned `pd.parquet_parquet`, which goes through `parquet.read_table/ParquetDataset`, and those paths will actually infer hive partitioning. But so not `ds.dataset(..)` ...

Reply:
That is confusing.  But you are correct:

```
import pyarrow as pa
import pyarrow.dataset as ds
import pyarrow.parquet as pq

table = pa.Table.from_pydict({'type': ['a', 'a', 'b', 'b'], 'value': [1, 2, 3, 4]})
ds.write_dataset(table, '/tmp/my_dataset', format='parquet', partitioning=['type'], partitioning_flavor='hive', existing_data_behavior='overwrite_or_ignore')

print(ds.dataset('/tmp/my_dataset').to_table().column_names)
print(ds.dataset('/tmp/my_dataset', partitioning='hive').to_table().column_names)
print(pq.read_table('/tmp/my_dataset').column_names)
```

Can we make this more intuitive somehow?  Maybe have `hive` be the default partitioning for a dataset similar to how it is for pq.read_table?

Reply:
I think "read with hive" is functionally an "auto" option.  If the write was directory partitioned it is harmless and no partitions are found.  If the write was hive partitioned it will detect it.

So I think we can change read without changing the write.  The only negative case will be the case where someone is using directory partitioning and their partition values actually have `=` inside of them.

Reply:
@AlenkaF and I ran into some similar issues again, reminding me that we should have better defaults (see also https://issues.apache.org/jira/browse/ARROW-15310 that I opened)

Agreed that for the read side, we can certainly already switch the default, as that is quite harmless. 

It would still be nice to have the write side to be consistent, but that's a bigger change. In theory we could do it with a deprecation cycle, though, now that we have the `partitioning_flavor` keyword. Currently that defaults to None, and you can set it to "hive" to get hive-style instead of directory partitioning. But so we could raise a warning if it is None (and a partitioning is specified), saying it will change to "hive" in the future, and people can specify "directory" or "hive" explicitly to silence the warning.

Reply:
Ok, I opened ARROW-15406 and ARROW-15407 which I think addresses what you are describing here.